src/main/scala/io/iohk/ethereum/vm/Memory.scala

```diff
@@ -0,0 +1,89 @@
+package io.iohk.ethereum.vm
+
+import akka.util.ByteString
+import cats.syntax.either._
+
+object Memory {
+
+  def empty: Memory = new Memory(ByteString())
+
+  private def zeros(size: Int): ByteString = ByteString(Array.fill[Byte](size)(0))
+
+}
+
+/**
+ * Volatile memory with 256 bit address space.
+ * Every mutating operation on a Memory returns a new updated copy of it.
+ *
+ * Related reading:
+ * https://solidity.readthedocs.io/en/latest/frequently-asked-questions.html#what-is-the-memory-keyword-what-does-it-do
+ * https://github.com/ethereum/go-ethereum/blob/master/core/vm/memory.go
+ */
+class Memory(private[vm] val underlying: ByteString) {
+
+  import Memory.zeros
+
+  def store(addr: DataWord, b: Byte): Memory = store(addr, ByteString(b))
+
+  def store(addr: DataWord, dw: DataWord): Memory = store(addr, dw.bytes)
+
+  def store(addr: DataWord, bytes: Array[Byte]): Memory = store(addr, ByteString(bytes))
+
+  /** Stores a ByteString under the given address.
+   * Underlying byte array is expanded if a ByteString doesn't fit into it.
+   * All empty cells of an expanded array are set to 0.
+   * This method may throw OOM.
+   */
+  def store(addr: DataWord, bs: ByteString): Memory = {
+    val idx: Int = addr.intValue
+    val newUnderlying: ByteString = if (idx + bs.length <= underlying.length) {
+      // a new buffer fits into an old buffer
+      val (prepending, following) = underlying.splitAt(idx)
+      prepending ++ bs ++ following.drop(bs.length)
+    } else if (idx <= underlying.length) {
+      // a new buffer partially fits into an old buffer
+      underlying.take(idx) ++ bs
+    } else {
+      // there is a gap (possibly empty) between an old buffer and a new buffer
+      underlying ++ zeros(idx - underlying.length) ++ bs
+    }
+    new Memory(newUnderlying)
+  }
+
+  def load(addr: DataWord): (DataWord, Memory) = {
+    doLoad(addr, DataWord.MaxLength) match {
+      case (bs, memory) => DataWord(bs) -> memory
+    }
+  }
+
+  def load(addr: DataWord, size: DataWord): (ByteString, Memory) = doLoad(addr, size.intValue)
+
+  /** Returns a ByteString of a given size starting at the given address of the Memory.
+   * Underlying byte array is expanded and filled with 0's if addr + size exceeds size
+   * of the memory.
+   * This method may throw OOM.
+   */
+  private def doLoad(addr: DataWord, size: Int): (ByteString, Memory) = {
```

Comment:
@LukasGasior1 the semantics of memory access in EVM are such that when you access a previously unallocated address the memory gets expanded accordingly and the caller pays the proportionate gas. In other words, loading from memory has the same effect on its size (and gas consumption) as storing in it.